### ISSUE TITLE
Cannot generate DoxygenLayout.xml

### ISSUE THREAD

**Issue by dimxy** [doxygen/doxygen#7328]
doxygen -l just creates a large 29M binary file.
version 1.8.16
I suspect it can't work on my locale (ru_RU)

**Comment by albert-github** [comment]
- Might be the problem but it is hard to tell. Can you attach the `DoxygenLayout.xml` in a compressed tar or zip file  in your question?
- Which OS are you using?
- I assume you gave the command `doxygen -v` to get the version information
- Did you try to generate the `DoxygenLayout.xml` on an empty directory?

**Comment by albert-github** [comment]
Please attach the file to the issue as indicated as external links as they might not be persistent.

**Comment by dimxy** [comment]
Thank you again,

- Sorry I did not get the first, I ran `doxygen -l`, how to run it to dump to the terminal? (I also tried to pass a filename for the layout file, the result was the same)

- 'doxygen -g' creates a correct Doxyfile

- yes, the generated doc is normally readable

**Comment by albert-github** [comment]
To dump to the terminal I expected that  `doxygen -l -` would work, but it doesn't ( :-( ). I should have tested it before. 
- Strangely enough with 1.8.16 it still writes the `DoxygenLayout.xml` 
- With the current master (560bd4f390b6f5ed3146ddd3def583bd074c35b9) doxygen even crashes, so something to check (for me).

I have no idea why the `DoxygenLayout.xml` file is not correct.

For your convenience I attached the default `DoxygenLayout.xml` of the 1.8.16 version. maybe you can use this file, as start, for your work ([DoxygenLayout.tar.gz](https://github.com/doxygen/doxygen/files/3735118/DoxygenLayout.tar.gz)). 
- can you also run a test where you place this file in an empty directory and see what happens when giving the command `doxygen -l`.

**Comment by dimxy** [comment]
Thank you,
I placed DoxygenLayout.xml into an empty dir and ran `doxygen -l`. It created a DoxygenLayout.xml.bak (7K) and also created the same DoxygenLayout.xml of 29M.

(BTW I tried it on Linux and there `doxygen -l` works fine, creates a good DoxygenLayout.xml)

**Comment by albert-github** [comment]
My conclusion at the moment is that it must be something strange with the settings on Windows.
Do you have the possibility to run the commands on another, independent, Windows 10 Rus system as well (to exclude the possibility that it is a problem on your system).

**Comment by dimxy** [comment]
unfortunately I don't have another Win 10 PC, 
I can try it on Windows server

**Comment by albert-github** [comment]
Always worthwhile to test it, is this also a system with "Windows 10 Rus"?

**Comment by dimxy** [comment]
No quite, it is a binary compatible platform but actually another OS

**Comment by albert-github** [comment]
Which OS is it? Anyway I think it will be a good test and know the results.
- did you build the executable yourself or did you use the distributed version?
- can you build the doxygen master version based on the github sources (so we can try do some debugging to see what happens)?
- the crash I talked about has been fixed (a minor cleanup has ben reverted).

**Comment by dimxy** [comment]
I just tried `doxygen -l` on another PC with Windows 7 SP1 Rus, with exactly the same result (29M file was generated).
In the case of Windows 10 I used the binary distrib with installer
in case of Windows 7 I used the binary distrib x64 without the installer

**Comment by dimxy** [comment]
Yes, okay, I'll try to build it from github master

**Comment by albert-github** [comment]
Thanks for the information, looks like, as we concluded before, it has to do something with the writing of the file when the locale is set to Russian.

I had a look in the code  for "locale" and saw some settings (don't know if these are the right values) but saw some `LC_...`.
Question: 
- are there any `LC_...` environment variables set? and when set to which values?
- is the environment variable `LANG` set? and when set to which values?

but as said before it is quite strange that only the `DoxygenLayout.xml` file is influenced by the problem.

**Comment by dimxy** [comment]
There is no any LC_.. or LANG variables in my environment (but evidently there are some local settings in Windows registry set by OS install)
I have built the doxygen binaries from github

**Comment by albert-github** [comment]
I had a look in my registry but didn't find here `LANG` orr `LC_...` except for `LC_RULES`, but I don't think the `LC_RULES` is of influence.
Did you check your registry for:
- `LC_` keys
- `LANG` keys as whole word

What happens when you do (I could not reproduce the problem for Russian though):
```
set LANG=en-US
doxygen -l
```

What happens when you add in the file `layout.cpp` just before the line: `  t << substitute(layout_default,"$doxygenversion",getVersion());` in the routine `void writeDefaultLayoutFile(const char *fileName)` (around line 1564):
```
  printf("%s\n",layout_default);
```
It should just dump the contents of the template used for the `DoxygenLayout.xml` to the terminal window.

**Comment by dimxy** [comment]
* If rebuild with that printf statment it outputs a visibly correct xml file.
* Setting LANG to `en-US` or `en_US` does not help (I think this variable is not used in Windows), but if I change the system language from `Russian` to `English(American)` in Windows Control Panel and reboot, doxygen generates a valid xml file of 7K.

**Comment by albert-github** [comment]
Some "good" results, thanks a lot for retesting with the system language set to `English(American)`.

Lets get some steps further and see which routine / function / class causes the problem.

Can you try in the same routine to replace:
```
  t << substitute(layout_default,"$doxygenversion",getVersion());
```
with
```
  printf("%s\n",getVersion());
  printf("%s\n",substitute(layout_default,"$doxygenversion",getVersion());
  t << layout_default;
```

What appears on the terminal (might actually hang your terminal / take a long time when 29MB is printed) and in the `DoxygenLayout.xml`?

**Comment by dimxy** [comment]
Seems I need to change the last statement to:
 `printf("%s\n", substitute(layout_default, "$doxygenversion", getVersion()).data());`
The output is:
1.8.17
<doxygenlayout version="1.0">
  
  
... skipped...
  </directory>                                                                                                                                                </doxygenlayout>          

But the DoxygenLayout.xml is still binary of 29M

**Comment by albert-github** [comment]
OK next tests, looks like we get somewhere.

What happens with `doxygen -e rtf -`, should give a normal readable test on the terminal, and `doxygen -e rtf ext.txt` should give a, readable, file of around 1567 bytes.

Can you try to replace:
```
  QTextStream t(&f);
  t << substitute(layout_default,"$doxygenversion",getVersion());
```
by
```
  QTextStream t(&f);
  t.setEncoding(QTextStream::UnicodeUTF8);
  t << substitute(layout_default,"$doxygenversion",getVersion());
```

**Comment by dimxy** [comment]
* Yes, `doxygen -e rtf -` outputs readable text and `doxygen -e rtf ext.txt` creates a file of 1567 bytes
* With the new changes the DoxygenLayout.xml file is created with normal xml content of 7665 bytes

**Comment by albert-github** [comment]
That is great news. With the rtfstyle.cpp I saw this construct and this brought me to the idea.
Thank you for you testing.

I've just pushed a proposed patch, pull request #7331

**Comment by dimxy** [comment]
Thank you too, I was glad to help!

**Comment by albert-github** [comment]
Code has been integrated in master on github (please don't close the issue as this will be done at the moment of a release).

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.8.17.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
